### PR TITLE
[Feature] Include Triton Flash attention implementation and optimizations for pipeline runtime 

### PR DESCRIPTION
**Pull request by zarzen**
## Description ##
This PR includes following modifications
- Including the triton flash attention op with dropout support, which is copied from the @jfc4050 https://github.com/jfc4050/fast-ops/ 
- Including the torch sdp operation for attention computation 
- Removed the data type conversion from `torch.Size` to `tuple`, it avoids sending the tensor size data as a list of integers in pipeline runtime 
- Removed the  `is_pipe_partitioned=False` and `is_grad_partitioned=False` flags to enable the partitioned activation/grad send recv in pipeline runtime. (previously, there is a hanging when this feature is enabled, it is now fixed)

## Checklist ##

- [x] PR's title starts with a category (e.g. [Bugfix], [Model], [Tutorial], etc)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented

### CODE REVIEW

File: examples/data_util.py [hunk 1]
```diff
@@ -82,9 +85,9 @@ def preprocessing_datasets(datasets, tokenizer, model_name, max_seq_length=1024)
     text_column_name = "text" if "text" in column_names else column_names[0]
 
     if tokenizer.model_max_length < max_seq_length:
-        raise ValueError(
+        logger.warn(
```
Comment (chhzh123) [inline examples/data_util.py:88]: These changes are covered by #107? I think a rebase can be useful to remove those changes.

File: examples/bert/deepspeed_hf.py [hunk 2]
```diff
@@ -198,7 +198,7 @@ def collate_fn(batch, enable_pipeline=False):
 
     train_loader, _ = get_dataloader(
         args.model_name,
-        "wikitext-103-v1",
+        "wikitext-2-v1",
```
Comment (chhzh123) [inline examples/bert/deepspeed_hf.py:201]: Why the dataset is changed?

File: examples/gpt2/run.sh [hunk 3]
```diff
@@ -4,4 +4,17 @@
 
 # This script is used to run GPT2 training with 3D parallelism enabled.
 # It is tested on a single AWS p3d instance with 8*V100 GPUs.
-deepspeed deepspeed_hf.py --batch_size 32 --micro_batch_size 4 --model_name gpt2-xl --iter_nums 20 --hidden-size 2048 --nlayers 24 --num-attn-heads 16 --dropout 0.1 --activation_function gelu --seq_len 2048 --pmp 2 --tmp 2 --checkpoint 1.0
\ No newline at end of file
+export CUDA_LAUNCH_BLOCKING=0
```
Comment (chhzh123) [inline examples/gpt2/run.sh:7]: Should blocking launch be removed?

File: slapo/framework_dialect/deepspeed/engine.py [hunk 4]
```diff
@@ -38,12 +38,4 @@ def init_ds_engine(model, **kwargs):
         mpu=mpu,
     )
 
-    # `is_pipe_partitioned=True` is the default value due to TP
```
Comment (chhzh123) [inline slapo/framework_dialect/deepspeed/engine.py:41]: This is no longer affected DS?

File: slapo/framework_dialect/deepspeed/pipeline.py [hunk 5]
```diff
@@ -297,12 +304,12 @@ def forward(self, *args, **kwargs):
...
-                assert (
-                    torch.is_tensor(unordered_args[idx])
-                    and unordered_args[idx].requires_grad is False
-                ), f"[{self.name}] The {i}-th argument {arg_name} is not a tensor or requires grad: {unordered_args[idx]}, which is not supported by DeepSpeed pipeline engine."
+            # if i > 0:
+            #     # https://github.com/microsoft/DeepSpeed/blob/v0.9.2/deepspeed/runtime/pipe/engine.py#L639
+            #     assert (
+            #         torch.is_tensor(unordered_args[idx])
+            #         and unordered_args[idx].requires_grad is False
+            #     ), f"[{self.name}] The {i}-th argument {arg_name} is not a tensor or requires grad: {unordered_args[idx]}, which is not supported by DeepSpeed pipeline engine."
```
Comment (chhzh123) [inline slapo/framework_dialect/deepspeed/pipeline.py:312]: Why these are commented?

File: slapo/op/flash_attention_triton.py [hunk 6]
```diff
@@ -0,0 +1,1032 @@
...
+import math
+from typing import Tuple
+
+import numpy as np
+import torch
+from torch import Tensor
+from torch.autograd import Function
+
+import triton
+import triton.language as tl
```
Comment (chhzh123) [inline slapo/op/flash_attention_triton.py:41]: Porting this file may require users to manually install Triton. Do we have error messages or instructions somewhere to inform users to install Triton?

**Comment by zarzen** [comment]
out dated
